File: _static/code/jv/jv.jl
#=

@author : Spencer Lyon <spencer.lyon@nyu.edu>

=#

using Distributions
using QuantEcon

# NOTE: only brute-force approach is available in bellman operator.
# Waiting on a simple constrained optimizer to be written in pure Julia

"""
A Jovanovic-type model of employment with on-the-job search.

The value function is given by

\\[V(x) = \\max_{ϕ, s} w(x, ϕ, s)\\]

for

    w(x, ϕ, s) := x(1 - ϕ - s) + β (1 - π(s)) V(G(x, ϕ)) +
                    β π(s) E V[ max(G(x, ϕ), U)

where

* `x`: human capital
* `s` : search effort
* `ϕ` : investment in human capital
* `π(s)` : probability of new offer given search level s
* `x(1 - ϕ - s)` : wage
* `G(x, ϕ)` : new human capital when current job retained
* `U` : Random variable with distribution F -- new draw of human capital

##### Fields

- `A::Real` : Parameter in human capital transition function
- `α::Real` : Parameter in human capital transition function
- `β::AbstractFloat` : Discount factor in (0, 1)
- `x_grid::AbstractVector` : Grid for potential levels of x
- `G::Function` : Transition `function` for human captial
- `π_func::Function` : `function` mapping search effort to
   the probability of getting a new job offer
- `F::UnivariateDistribution` : A univariate distribution from which
   the value of new job offers is drawn
- `quad_nodes::Vector` : Quadrature nodes for integrating over ϕ
- `quad_weights::Vector` : Quadrature weights for integrating over ϕ
- `ϵ::AbstractFloat` : A small number, used in optimization routine

"""
struct JvWorker{TR <: Real,
                TF <: AbstractFloat,
                TUD <: UnivariateDistribution,
                TAV <: AbstractVector,
                TV <: Vector}
    A::TR
    α::TR
    β::TF
    x_grid::TAV
    G::Function
    π_func::Function
    F::TUD
    quad_nodes::TV
    quad_weights::TV
    ϵ::TF

end

"""
Constructor with default values for `JvWorker`

##### Arguments

 - `A::Real(1.4)` : Parameter in human capital transition function
 - `α::Real(0.6)` : Parameter in human capital transition function
 - `β::Real(0.96)` : Discount factor in (0, 1)
 - `grid_size::Integer(50)` : Number of points in discrete grid for `x`
 - `ϵ::Float(1e-4)` : A small number, used in optimization routine

##### Notes

There is also a version of this function that accepts keyword arguments for
each parameter

"""
# use key word argument
function JvWorker(;A::Real=1.4,
                   α::Real=0.6,
                   β::Real=0.96,
                   grid_size::Integer=50,
                   ϵ::AbstractFloat=1e-4)

    G(x, ϕ) = A .* (x .* ϕ).^α
    π_func = sqrt
    F = Beta(2, 2)

    # integration bounds
    a, b = quantile(F, 0.005), quantile(F, 0.995)

    # quadrature nodes/weights
    nodes, weights = qnwlege(21, a, b)

    # Set up grid over the state space for DP
    # Max of grid is the max of a large quantile value for F and the
    # fixed point y = G(y, 1).
    grid_max = max(A^(1.0 / (1.0 - α)), quantile(F, 1 - ϵ))

    # range for linspace(ϵ, grid_max, grid_size). Needed for
    # CoordInterpGrid below
    x_grid = linspace(ϵ, grid_max, grid_size)

    JvWorker(A, α, β, x_grid, G, π_func, F, nodes, weights, ϵ)
end


"""
Apply the Bellman operator for a given model and initial value,
returning only the value function

##### Arguments

- `jv::JvWorker` : Instance of `JvWorker`
- `V::Vector`: Current guess for the value function
- `new_V::Vector` : Storage for updated value function

##### Returns

None, `new_V` is updated in place with the value function.

##### Notes

Currently, only the brute-force approach is available.
We are waiting on a simple constrained optimizer to be written in pure Julia

"""
function bellman_operator!(jv::JvWorker, V::AbstractVector, new_V::AbstractVector)

    # simplify notation
    G, π_func, F, β, ϵ = jv.G, jv.π_func, jv.F, jv.β, jv.ϵ
    nodes, weights = jv.quad_nodes, jv.quad_weights

    # prepare interpoland of value function
    Vf = LinInterp(jv.x_grid, V)

    # instantiate the linesearch variables
    max_val = -1.0
    cur_val = 0.0
    max_s = 1.0
    max_ϕ = 1.0
    search_grid = linspace(ϵ, 1.0, 15)

    for (i, x) in enumerate(jv.x_grid)

        function w(z)
            s, ϕ = z
            h(u) = [Vf(max(G(x, ϕ), uval)) * pdf(F, uval) for uval in u]
            integral = do_quad(h, nodes, weights)
            q = π_func(s) * integral + (1.0 - π_func(s)) * Vf(G(x, ϕ))

            return - x * (1.0 - ϕ - s) - β * q
        end

        for s in search_grid
            for ϕ in search_grid
                cur_val = ifelse(s + ϕ <= 1.0, -w((s, ϕ)), -1.0)
                if cur_val > max_val
                    max_val, max_s, max_ϕ = cur_val, s, ϕ
                end
            end
        end

        new_V[i] = max_val
    end
end

"""
Apply the Bellman operator for a given model and initial value, returning policies

##### Arguments

- `jv::JvWorker` : Instance of `JvWorker`
- `V::Vector`: Current guess for the value function
- `out::Tuple{Vector, Vector}` : Storage for the two policy rules

##### Returns

None, `out` is updated in place with the two policy functions.

##### Notes

Currently, only the brute-force approach is available.
We are waiting on a simple constrained optimizer to be written in pure Julia

"""
function bellman_operator!(jv::JvWorker,
                           V::AbstractVector,
                           out::Tuple{AbstractVector, AbstractVector})

    # simplify notation
    G, π_func, F, β, ϵ = jv.G, jv.π_func, jv.F, jv.β, jv.ϵ
    nodes, weights = jv.quad_nodes, jv.quad_weights

    # prepare interpoland of value function
    Vf = LinInterp(jv.x_grid, V)

    # instantiate variables
    s_policy, ϕ_policy = out[1], out[2]

    # instantiate the linesearch variables
    max_val = -1.0
    cur_val = 0.0
    max_s = 1.0
    max_ϕ = 1.0
    search_grid = linspace(ϵ, 1.0, 15)

    for (i, x) in enumerate(jv.x_grid)

        function w(z)
            s, ϕ = z
            h(u) = [Vf(max(G(x, ϕ), uval)) * pdf(F, uval) for uval in u]
            integral = do_quad(h, nodes, weights)
            q = π_func(s) * integral + (1.0 - π_func(s)) * Vf(G(x, ϕ))

            return - x * (1.0 - ϕ - s) - β * q
        end

        for s in search_grid
            for ϕ in search_grid
                cur_val = ifelse(s + ϕ <= 1.0, -w((s, ϕ)), -1.0)
                if cur_val > max_val
                    max_val, max_s, max_ϕ = cur_val, s, ϕ
                end
            end
        end

      s_policy[i], ϕ_policy[i] = max_s, max_ϕ
  end
end

function bellman_operator(jv::JvWorker, V::AbstractVector; ret_policies::Bool=false)
    out = ifelse(ret_policies, (similar(V), similar(V)), similar(V))
    bellman_operator!(jv, V, out)
    return out
end

File: rst_files/_static/code/jv/jv.py
import numpy as np
from scipy.integrate import fixed_quad as integrate
from scipy.optimize import minimize
import scipy.stats as stats

ϵ = 1e-4  # A small number, used in the optimization routine

class JvWorker:
    r"""
    A Jovanovic-type model of employment with on-the-job search. The
    value function is given by

    .. math::

        V(x) = \max_{ϕ, s} w(x, ϕ, s)

    for

    .. math::

        w(x, ϕ, s) := x(1 - ϕ - s)
                        + β (1 - π(s)) V(G(x, ϕ))
                        + β π(s) E V[ \max(G(x, ϕ), U)]

    Here

    * x = human capital
    * s = search effort
    * :math:`ϕ` = investment in human capital
    * :math:`π(s)` = probability of new offer given search level s
    * :math:`x(1 - ϕ - s)` = wage
    * :math:`G(x, ϕ)` = new human capital when current job retained
    * U = RV with distribution F -- new draw of human capital

    Parameters
    ----------
    A : scalar(float), optional(default=1.4)
        Parameter in human capital transition function
    α : scalar(float), optional(default=0.6)
        Parameter in human capital transition function
    β : scalar(float), optional(default=0.96)
        Discount factor
    grid_size : scalar(int), optional(default=50)
        Grid size for discretization
    G : function, optional(default=lambda x, ϕ: A * (x * ϕ)**α)
        Transition function for human captial
    π : function, optional(default=sqrt)
        Function mapping search effort (:math:`s \in (0,1)`) to
        probability of getting new job offer
    F : distribution, optional(default=Beta(2,2))
        Distribution from which the value of new job offers is drawn

    Attributes
    ----------
    A, α, β : see Parameters
    x_grid : array_like(float)
        The grid over the human capital

    """

    def __init__(self, A=1.4, α=0.6, β=0.96, grid_size=50,
                 G=None, π=np.sqrt, F=stats.beta(2, 2)):
        self.A, self.α, self.β = A, α, β

        # === set defaults for G, π and F === #
        self.G = G if G is not None else lambda x, ϕ: A * (x * ϕ)**α
        self.π = π
        self.F = F

        # === Set up grid over the state space for DP === #
        # Max of grid is the max of a large quantile value for F and the
        # fixed point y = G(y, 1).
        grid_max = max(A**(1 / (1 - α)), self.F.ppf(1 - ϵ))
        self.x_grid = np.linspace(ϵ, grid_max, grid_size)


    def bellman_operator(self, V, brute_force=False, return_policies=False):
        """
        Returns the approximate value function TV by applying the
        Bellman operator associated with the model to the function V.

        Returns TV, or the V-greedy policies s_policy and ϕ_policy when
        return_policies=True.  In the function, the array V is replaced below
        with a function Vf that implements linear interpolation over the
        points (V(x), x) for x in x_grid.


        Parameters
        ----------
        V : array_like(float)
            Array representing an approximate value function
        brute_force : bool, optional(default=False)
            Default is False. If the brute_force flag is True, then grid
            search is performed at each maximization step.
        return_policies : bool, optional(default=False)
            Indicates whether to return just the updated value function
            TV or both the greedy policy computed from V and TV


        Returns
        -------
        s_policy : array_like(float)
            The greedy policy computed from V.  Only returned if
            return_policies == True
        new_V : array_like(float)
            The updated value function Tv, as an array representing the
            values TV(x) over x in x_grid.

        """
        # === simplify names, set up arrays, etc. === #
        G, π, F, β = self.G, self.π, self.F, self.β
        Vf = lambda x: np.interp(x, self.x_grid, V)
        N = len(self.x_grid)
        new_V, s_policy, ϕ_policy = np.empty(N), np.empty(N), np.empty(N)
        a, b = F.ppf(0.005), F.ppf(0.995)           # Quantiles, for integration
        c1 = lambda z: 1.0 - sum(z)                 # used to enforce s + ϕ <= 1
        c2 = lambda z: z[0] - ϵ                     # used to enforce s >= ϵ
        c3 = lambda z: z[1] - ϵ                     # used to enforce ϕ >= ϵ
        guess = (0.2, 0.2)
        constraints = [{"type": "ineq", "fun": i} for i in [c1, c2, c3]]

        # === solve r.h.s. of Bellman equation === #
        for i, x in enumerate(self.x_grid):

            # === set up objective function === #
            def w(z):
                s, ϕ = z
                h = lambda u: Vf(np.maximum(G(x, ϕ), u)) * F.pdf(u)
                integral, err = integrate(h, a, b)
                q = π(s) * integral + (1.0 - π(s)) * Vf(G(x, ϕ))
                # == minus because we minimize == #
                return - x * (1.0 - ϕ - s) - β * q

            # === either use SciPy solver === #
            if not brute_force:
                max_s, max_ϕ = minimize(w, guess, constraints=constraints,
                                        options={"disp": 0},
                                        method="COBYLA")["x"]
                max_val = -w((max_s, max_ϕ))

            # === or search on a grid === #
            else:
                search_grid = np.linspace(ϵ, 1.0, 15)
                max_val = -1.0
                for s in search_grid:
                    for ϕ in search_grid:
                        current_val = -w((s, ϕ)) if s + ϕ <= 1.0 else -1.0
                        if current_val > max_val:
                            max_val, max_s, max_ϕ = current_val, s, ϕ

            # === store results === #
            new_V[i] = max_val
            s_policy[i], ϕ_policy[i] = max_s, max_ϕ

        if return_policies:
            return s_policy, ϕ_policy
        else:
            return new_V
